feature/platform/ate_tests/telemetry_inventory_test/telemetry_inventory_test.go

```diff
@@ -426,228 +443,234 @@ func TestTempSensor(t *testing.T) {
-				t.Logf("Optics is detected in %s with expected parent: %s", cName, card.GetParent())
-			} else {
-				t.Logf("Optics is not installed in %s, skip testing this transceiver", cName)
-				continue
+				validCards = append(validCards, card)
 			}
 		}
-
-		if p.descriptionValidation {
-			description := card.GetDescription()
-			t.Logf("Component %s Description: %s", cName, description)
-			if description == "" {
-				t.Errorf("Component %s Description: got empty string, want non-empty string", cName)
-			}
+	default:
+		validCards = cards
+	}
+	for _, card := range validCards {
+		if card.Name == nil {
+			t.Errorf("Encountered a component with no Name")
+			continue
 		}
+		cName := card.GetName()
+		t.Run(cName, func(t *testing.T) {
+			if p.descriptionValidation {
+				t.Logf("Component %s Description: %s", cName, card.GetDescription())
+				if card.GetDescription() == "" {
+					t.Errorf("Component %s Description: got empty string, want non-empty string", cName)
+				}
+			}
 
-		if p.idValidation {
-			if deviations.SwitchChipIDUnsupported(dut) {
-				t.Logf("Skipping check for Id due to deviation SwitChipIDUnsupported")
-			} else {
-				id := card.GetId()
-				t.Logf("Component %s Id: %s", cName, id)
-				if id == "" {
-					t.Errorf("Component %s Id: got empty string, want non-empty string", cName)
+			if p.idValidation {
+				if deviations.SwitchChipIDUnsupported(dut) {
+					t.Logf("Skipping check for Id due to deviation SwitChipIDUnsupported")
+				} else {
+					id := card.GetId()
+					t.Logf("Component %s Id: %s", cName, id)
+					if id == "" {
+						t.Errorf("Component %s Id: got empty string, want non-empty string", cName)
+					}
 				}
 			}
-		}
 
-		if p.nameValidation {
-			name := card.GetName()
-			t.Logf("Component %s Name: %s", cName, name)
-			if name == "" {
-				t.Errorf("Encountered empty Name for component %s", cName)
+			if p.nameValidation {
+				name := card.GetName()
+				t.Logf("Component %s Name: %s", cName, name)
+				if name == "" {
+					t.Errorf("Encountered empty Name for component %s", cName)
+				}
 			}
-		}
 
-		if p.partNoValidation {
-			partNo := card.PartNo
-			t.Logf("Component %s PartNo: %v", cName, partNo)
-			if partNo == nil {
-				if card.GetType() == fanType {
-					fanTray := card.GetParent()
-					fanTrayPartNo := gnmi.Lookup(t, dut, gnmi.OC().Component(fanTray).PartNo().State())
-					fanTrayPartNoVal, ftpnPresent := fanTrayPartNo.Val()
-					if !ftpnPresent {
-						t.Errorf("PartNo for Component %s and its parent: got empty string, want non-empty string", cName)
+			if p.partNoValidation {
+				partNo := card.PartNo
+				t.Logf("Component %s PartNo: %v", cName, partNo)
+				if partNo == nil {
+					switch card.GetType() {
+					case componentType["Fan"]:
+						fallthrough
+					case componentType["Storage"]:
+						fallthrough
+					case componentType["Cpu"]:
+						parent := card.GetParent()
+						ppn, ok := gnmi.Lookup(t, dut, gnmi.OC().Component(parent).PartNo().State()).Val()
```

Comment:
In the description of the PR it says:
"While it is not required for STORAGE and CPU components to have their own PartNo and SerialNo, we do expect that, if we look at their parent component (or parent of parent component), we would eventually reach the CHASSIS component which must have a PartNo and SerialNo."

However in the implementation it seems to be making assumptions as to which parts do or don't contain a part number. I think this should be refactored to detect if the serial/part-number exists, and if not search all parents until an ancestor is found that contains these fields. 

This means that perhaps you end up with the linecard, or perhaps the chassis, but you are also most likely to find the component which you would need to RMA.

Reply:
Right, currently this code doesn't check for the first FRU. e.g a CPU's direct parent may not be a FRU, a grandparent however might be. But the current code will fail in such a scenario since it will pass only when the direct parent is a FRU.

Reply:
Done!